File: apps/web/src/types/itinerary-version.ts
import type { Itinerary } from "@/types/trip";

export type ItineraryVersionSource =
  | "GENERATED"
  | "MANUAL_EDIT"
  | "REGENERATE_DAY"
  | "REGENERATE_ITEM"
  | "RESTORED";

export type ItineraryVersionSummary = {
  id: string;
  tripId: string;
  versionNumber: number;
  source: ItineraryVersionSource;
  metadata?: Record<string, unknown> | null;
  createdAt: string;
};

export type ItineraryVersionDetail = ItineraryVersionSummary & {
  itinerary: Itinerary;
};

export type ListItineraryVersionsResponse = {
  items: ItineraryVersionSummary[];
  limit: number;
  offset: number;
};

File: services/trip-service/internal/domain/entity/itinerary_version.go
package entity

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
)

// ItineraryVersionSource identifies the operation that produced a snapshot.
type ItineraryVersionSource string

const (
	ItineraryVersionSourceGenerated                 ItineraryVersionSource = "GENERATED"
	ItineraryVersionSourceManualEdit                ItineraryVersionSource = "MANUAL_EDIT"
	ItineraryVersionSourceRegenerateDay             ItineraryVersionSource = "REGENERATE_DAY"
	ItineraryVersionSourceRegenerateItem            ItineraryVersionSource = "REGENERATE_ITEM"
	ItineraryVersionSourceBudgetOptimizationApplied ItineraryVersionSource = "BUDGET_OPTIMIZATION_APPLIED"
	ItineraryVersionSourceAIPolicyRepairApplied     ItineraryVersionSource = "AI_POLICY_REPAIR"
	ItineraryVersionSourceCostSplitUpdated          ItineraryVersionSource = "COST_SPLIT_UPDATED"
	ItineraryVersionSourceRestored                  ItineraryVersionSource = "RESTORED"
	ItineraryVersionSourceCreatedFromTemplate       ItineraryVersionSource = "CREATED_FROM_TEMPLATE"
	ItineraryVersionSourceCreatedFromTemplateAI     ItineraryVersionSource = "CREATED_FROM_TEMPLATE_AI"
)

// ItineraryVersion is a full JSONB snapshot of a trip itinerary at one point in
// that trip's edit history. The itinerary is intentionally stored as raw JSON so
// snapshots can preserve the exact API shape accepted by itinerary editing v1.
type ItineraryVersion struct {
	ID              uuid.UUID
	TripID          uuid.UUID
	UserID          uuid.UUID
	CreatedByUserID *uuid.UUID
	VersionNumber   int
	Source          ItineraryVersionSource
	Itinerary       json.RawMessage
	Metadata        map[string]any
	CreatedAt       time.Time
}
